common/OpTestSSH.py

```diff
@@ -0,0 +1,180 @@
+            self.console.terminate()
+            self.state = ConsoleState.DISCONNECTED
+
+        cmd = ("sshpass -p %s " % (self.password)
+               + " ssh -q"
+               + " -p %s" % str(self.port)
+               + " -l %s %s" % (self.username, self.host)
+               + " -o'RSAAuthentication=no' -o 'PubkeyAuthentication=no'"
+               + " -o 'UserKnownHostsFile=/dev/null' "
+               + " -o 'StrictHostKeyChecking=no'"
+           )
+
+        print cmd
+        consoleChild = OPexpect.spawn(cmd,logfile=self.logfile,
+                failure_callback=set_system_to_UNKNOWN,
+                failure_callback_data=self.system)
+        self.state = ConsoleState.CONNECTED
+        self.console = consoleChild
+        # Users expecting "Host IPMI" will reference console.sol so make it available
+        self.sol = self.console
+        return consoleChild
+
+    def get_console(self):
+        if self.state == ConsoleState.DISCONNECTED:
+            self.connect()
+
+        count = 0
+        while (not self.console.isalive()):
+            print '# Reconnecting'
+            if (count > 0):
+                time.sleep(1)
+            self.connect()
+            count += 1
+            if count > 120:
+                raise "SSH: not able to get console"
+
+        return self.console
+
+    def run_command(self, command, timeout=60):
+        console = self.get_console()
+
+        if self.prompt:
+            prompt = self.prompt
+        else:
+            console.sendline('PS1=[console-pexpect]\#')
+            console.expect("\n") # from us, because echo
+            prompt = "\[console-pexpect\]#$"
+
+        rc = console.expect([prompt], timeout)
+        output = console.before
+
+        console.sendline(command)
+        console.expect("\n") # from us
+        rc = None
+        output = None
+        exitcode = None
+        try:
+            rc = console.expect([prompt], timeout)
+            output = console.before
+            console.sendline("echo $?")
+            console.expect("\n") # from us
+            rc = console.expect([prompt], timeout)
+            exitcode = int(console.before)
+        except pexpect.TIMEOUT as e:
+            print e
+            print "# TIMEOUT waiting for command to finish."
+            print "# Attempting to control-c"
+            try:
+                console.sendcontrol('c')
+                rc = console.expect(["\[console-pexpect\]#$"], 10)
+                if rc == 0:
+                    raise CommandFailed(command, "TIMEOUT", -1)
+            except pexpect.TIMEOUT:
+                print "# Timeout trying to kill timed-out command."
+                print "# Failing current command and attempting to continue"
+                self.terminate()
+                raise CommandFailed("ssh -p " + self.port, "timeout", -1)
+            raise e
+
+        if rc == 0:
+            res = output
+            res = res.splitlines()
+            self.terminate()
```

Comment:
why are we terminating here, we want both the console and ssh session to be active through entire test. Otherwise for each command we need to connect/terminate.

Reply:
On longer tests such as flashing a host PNOR I was seeing timeouts if we didn't terminate the connection - however this may have been a roundabout side affect of the above problem. Will double check this case.

Reply:
For this we can add `self.pxssh.isalive()` in run_command API to check if the console is alive, if not please call the function get_console() after setting SSH state to Disconnected.

Reply:
We will follow this procedure for complete Host/SSH API'a.
1. Get the console at starting of test.
2. Login and Set unique prompt once if host is UP.
3. Run the commands.( We will always check console is alive before running commands, if not re-connect it).